Template lambda for scalar calls

Yurlungur opened the pull request:
<!--Provide a general summary of your changes in the title above, for
example "fix bug in ideal gas EOS.".  Please avoid
non-descriptive titles such as "Addresses issue #8576".-->

## PR Summary

The work @gopsub is doing requires passing in many mesh-sized arrays into the lambda parameter. Currently on a per-cell basis, this requires packing those arrays into local contiguous arrays for lambdas. This can be cumbersome and expensive, especially for EOSs where that data is unneeded and is just a no-op.

This PR adds indexers/accessors to the lambda parameter for scalar calls, meaning data can now be passed in in a layout agnostic way. This should also reduce pain points in phoebus/riot, where the lambda arrays are used regularly and currently require contiguous accesses.

<!--Please provide at least 1-2 sentences describing the pull request in
detail.  Why is this change required?  What problem does it solve?-->

<!--If it fixes an open issue, please link to the issue here.-->

## PR Checklist

<!-- Note that some of these check boxes may not apply to all pull requests -->

- [x] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Format your changes by using the `make format` command after configuring with `cmake`.
- [x] Document any new features, update documentation for changes made.
- [x] Make sure the copyright notice on any files you modified is up to date.
- [x] After creating a pull request, note it in the CHANGELOG.md file.
- [ ] LANL employees: make sure tests pass both on the github CI and on the Darwin CI

If preparing for a new release, in addition please check the following:
- [ ] Update the version in cmake.
- [ ] Move the changes in the CHANGELOG.md file under a new header for the new release, and reset the categories.

Yurlungur commented on singularity-eos/base/variadic_utils.hpp:
```diff
@@ -16,13 +16,28 @@
 // Some generic variatic utilities
 // ======================================================================
 
+// SFINAE to check if a value is a null ptr
+template <typename T, typename = typename std::enable_if<
+                          std::is_pointer<typename std::remove_reference<
+                              typename std::remove_cv<T>::type>::type>::value>::type>
+constexpr inline bool is_nullptr(T &&t) {
+  return std::forward<T>(t) == nullptr;
+}
+template <typename T,
+          typename std::enable_if<!std::is_pointer<typename std::remove_reference<
+              typename std::remove_cv<T>::type>::type>::value>::type * = nullptr>
+constexpr inline bool is_nullptr(T &&) {
+  return false;
+}
```
This can be cleaned up when we move to C++17. Thanks to @dholladay00 for the particular version of this, which checks edge cases.

jhp-lanl replied:
It's too late on a Friday for understanding the nestedness going on with `::type>::type>::value>::type>` 🤯

Yurlungur replied:
Haha blame/thank @dholladay00 for that one. Conceptually all it's doing is checking whether or not T is a pointer. If T is a pointer, check if it's null. Otherwise, return false.

All the other nested type nonsense is checking for edge cases, references to pointers, const pointers, etc.

Yurlungur replied:
The C++17 version will need to retain the nested templates, but will be able to move it out of `std::enable_if` and into `if constexpr`

jhp-lanl replied:
This is something we did in EAP to handle this. It might make sense to implement this to just make things slightly more readable and reusable if we need to do this elsewhere:
```c++
// C++14 implementation of std::remove_cvref (availalbe since C++20)
template<typename T>
struct remove_cvref { typedef std::remove_cv_t<std::remove_reference_t<T>> type; };

// Helper types
template<typename T>
using remove_cvref_t = typename remove_cvref<T>::type;
```

Credit to Diego via CJ for our implementation

Yurlungur replied:
:+1: Done.

Yurlungur commented:
@rbberger please help with the pybind here. We need to force the python module to specialize the lambdas to `Real*`

rbberger commented:
@Yurlungur turned out to be a trivial change.

Yurlungur commented:
This is now ready for review. @dholladay00 @gopsub @jhp-lanl

Yurlungur commented:
Ping @jhp-lanl @dholladay00 reminder to review.

jhp-lanl reviewed (approved):
Looks good to me. I appreciate the `is_nullptr` function. I think it's _very_ clear now what's going on and if I want to wade into the template nonsense I can look at the `variadic_utils.hpp` file 🙂

dholladay00 commented on singularity-eos/eos/eos_spiner.hpp:
```diff
@@ -102,56 +103,73 @@ class SpinerEOSDependsRhoT : public EosBase<SpinerEOSDependsRhoT> {
+  template <typename Indexer_t = Real *>
+  PORTABLE_INLINE_FUNCTION Real GruneisenParamFromDensityInternalEnergy(
+      const Real rho, const Real sie,
+      Indexer_t &&lambda = static_cast<Real *>(nullptr)) const;
```
should all Eos's get these static casts for the default nullptr?

Yurlungur replied:
It only really matters for the ones I did it for. But I'll change it everywhere for consistency.

Yurlungur commented on singularity-eos/closure/mixed_cell_models.hpp:
```diff
@@ -566,7 +566,7 @@ class PTESolverRhoT : public mix_impl::PTESolverBase<EOSIndexer, RealIndexer> {
     // TODO(JCD): use whatever lambdas are passed in
     /*for (int m = 0; m < nmat; m++) {
-      if (lambda[m] != nullptr) Cache[m] = lambda[m];
+      if (variadic_utils::is_nullptr(lambda[m])) Cache[m] = lambda[m];
```
Oops good catch. This line is commented out so it doesn't actually matter. But updated.

Yurlungur commented:
@dholladay00 I think I addressed all your comments. Can you re-review and approve if you're happy?

dholladay00 commented:
> @dholladay00 I think I addressed all your comments. Can you re-review and approve if you're happy?

Is this passing gitlab CI @Yurlungur ?

Yurlungur commented:
> > @dholladay00 I think I addressed all your comments. Can you re-review and approve if you're happy?
> 
> Is this passing gitlab CI @Yurlungur ?

Running. Won't merge until it passes.

mauneyc-LANL commented on python/module.hpp:
```diff
@@ -27,14 +27,14 @@ using namespace singularity;
+template<typename T, PORTABLE_FUNCTION Real(T::*Func)(const Real, const Real, Real*&&) const>
 Real two_params_no_lambda(const T& self, const Real a, const Real b) {
-  return (self.*Func)(a, b, nullptr);
+  return (self.*Func)(a, b, static_cast<Real*>(nullptr));
```
Can this be simplified?
```c++
// somewhere above in scope
constexpr auto realnp(){ return static_cast<Real*>(nullptr); }
// then e.g. 
return (self.*Func)(a, b, realnp());
```
Style preference here, so if not wanted then it's fine as is.

Yurlungur replied:
Good suggestion :+1: I'll switch this at least in the explicit calls.

Yurlungur commented on singularity-eos/base/variadic_utils.hpp:
```diff
@@ -16,13 +16,37 @@
+// SFINAE to check if a value is a null ptr
+template <typename T, typename = typename std::enable_if<
+                          std::is_pointer<remove_cvref_t<T>>::value>::type>
+inline bool is_nullptr(T &&t) {
```
Good catch---they actually need to be for device.

Yurlungur commented:
Addressed @mauneyc-LANL 's comments. As soon as tests pass on re-git I will merge.

Yurlungur commented:
Commenting here for posterity that when we switch to C++17 we will be able to do the following, which will allow us to properly handle passing in `nullptr`:

```C++
using Real = double;

template <typename T>
struct remove_cvref {
  typedef std::remove_cv_t<std::remove_reference_t<T>> type;
};
template <typename T>
using remove_cvref_t = typename remove_cvref<T>::type;

template<typename T>
constexpr inline bool is_null(T &&t) {
    if constexpr(std::is_pointer<remove_cvref_t<T>>::value) {
        return std::forward<T>(t) == nullptr;
    } else {
        return false;
    }
}

#define SG_DO_IF_NOT_NULL(lambda) \
  if constexpr (!std::is_null_pointer<remove_cvref_t<decltype(lambda)>>::value) \
    if (!is_null(lambda)) 

template<typename T>
void DoSomething(T &&t) {
    SG_DO_IF_NOT_NULL(t) {
        std:: cout << t[0] << std::endl;
    }
}

int main() {

    auto q = nullptr;
    std::vector<Real> v = {1, 2, 3, 4};
    DoSomething(v);
    DoSomething(q);

    return 0;
}
```

Yurlungur commented:
Tests pass on re-git. Merging.
